fix: check overlapping paths separately for directories and contents

Zebradil opened the pull request:
This PR fixes the logic of the overlapping paths check. In the original versions, paths of directories were joined with paths of their contents before the check. This allowed the following configuration to be accepted:

```yaml
apiVersion: vendir.k14s.io/v1alpha1
kind: Config
directories:
  - path: charts
    contents:
      - path: foo
        inline:
          paths:
            foo.txt: file-content
  - path: charts
    contents:
      - path: bar
        inline:
          paths:
            bar.txt: file-content
```

The result of `vendir sync` would be:

```
❯ vendir sync
Fetching: charts + foo (inline)

Fetching: charts + bar (inline)

Lock config

apiVersion: vendir.k14s.io/v1alpha1
directories:
- contents:
  - inline: {}
    path: foo
  path: charts
- contents:
  - inline: {}
    path: bar
  path: charts
kind: LockConfig

Succeeded

❯ exa -T
.
├── charts
│  └── bar
│     └── bar.txt
├── vendir.lock.yml
└── vendir.yml
```

One of the contents is overwritten by the other.

With this PR, having the same paths for two directories will result in an error.

Zebradil committed:
fix: check overlapping paths separately for directories and contents

Signed-off-by: German Lashevich <german.lashevich@gmail.com>

Zebradil committed:
fix: do not split directory into multiple when filtering paths

Signed-off-by: German Lashevich <german.lashevich@gmail.com>

kumaritanushree commented:
@Zebradil I think you are trying to do the same thing what @fritzduchardt has done in this PR https://github.com/carvel-dev/vendir/pull/325 

I have added a solution as well there if that fulfil your requirements. Else we will have more discussion on this in next community meeting to have a final decision wether we want this change or not.
Thanks

Zebradil commented:
Hi @kumaritanushree, thank you for looking into this.

I saw #325 before creating this PR, but my proposal differs. The PR from @fritzduchardt proposes to *allow* the same paths to be managed by different `directory` entities in the vendir configuration. My PR proposes to *disallow* duplicated paths.

I've been using vendir for more than three years and I always was sure that one path could be managed only by one `directory`. I can be wrong about that. Maybe @cppforlife can share the initial idea behind this.

joaopapereira commented:
Let us talk about this in tomorrow's community meeting @fritzduchardt @Zebradil if yll can attend

Zebradil committed:
refactor: revert changes to grouping of imports

Signed-off-by: German Lashevich <german.lashevich@gmail.com>

Zebradil committed:
test: add e2e tests for overlapping paths

Signed-off-by: German Lashevich <german.lashevich@gmail.com>

Zebradil committed:
chore: remove overlapping-dir example as unused

Signed-off-by: German Lashevich <german.lashevich@gmail.com>

Zebradil commented:
@joaopapereira as we agreed in the community meeting, I went ahead and:

- replaced the old e2e test for overlapping paths with the new one with more test cases,
- reverted ordering changes to the imports block,
- removed the overlapping-dir example (it was used in the old e2e test).

I'm unsure about the last point, but I don't see much value in having an example of _incorrect_ configuration (I bet, code copilots learn from it 😄).

kumaritanushree commented:
Changes looks good to me. Tested, error message also looks good.

joaopapereira reviewed (approved):
LGTM
